src/kvstore/raftex/Host.cpp

```diff
@@ -270,11 +270,29 @@ Host::prepareAppendLogRequest() {
   VLOG(2) << idStr_ << "Prepare AppendLogs request from Log " << lastLogIdSent_ + 1 << " to "
           << logIdToSend_;
 
+  auto makeReq = [this]() -> std::shared_ptr<cpp2::AppendLogRequest> {
+    auto req = std::make_shared<cpp2::AppendLogRequest>();
+    req->space_ref() = part_->spaceId();
+    req->part_ref() = part_->partitionId();
+    req->current_term_ref() = logTermToSend_;
+    req->committed_log_id_ref() = committedLogId_;
+    req->leader_addr_ref() = part_->address().host;
+    req->leader_port_ref() = part_->address().port;
+    req->last_log_term_sent_ref() = lastLogTermSent_;
+    req->last_log_id_sent_ref() = lastLogIdSent_;
+    return req;
+  };
+
   // We need to use lastLogIdSent_ + 1 to check whether need to send snapshot
   if (UNLIKELY(lastLogIdSent_ + 1 < part_->wal()->firstLogId())) {
     return startSendSnapshot();
   }
 
+  if (lastLogIdSent_ == logIdToSend_) {
+    auto req = makeReq();
+    return req;
```

Comment:
I am very curious, why do you need to use `AppendLogRequest` to send empty data, is it a waste of rpc?
Do not send AppendLogRequest, or return directly in RaftPart::replicateLogs?

Reply:
> I am very curious, why do you need to use `AppendLogRequest` to send empty data, is it a waste of rpc? Do not send AppendLogRequest, or return directly in RaftPart::replicateLogs?
if change leader and change back, the remote's logs may rollback, we must confirm the remote host really don't need logs.